client/cli/src/lib.rs

```diff
@@ -177,461 +171,397 @@ fn is_node_name_valid(_name: &str) -> Result<(), &str> {
-		builder: F,
-		spec_factory: S,
-		exit: Exit,
-	) -> error::Result<()>
-	where S: FnOnce(&str) -> Result<Option<ChainSpec<G, E>>, String>,
-		F: FnOnce(Configuration<C, G, E>) -> Result<B, error::Error>,
-		B: ServiceBuilderCommand,
-		C: Default,
-		G: RuntimeGenesis,
-		E: ChainSpecExtension,
-		Exit: IntoExit
-	{
-		let mut config = create_config_with_db_path(spec_factory, &self.params.shared_params, self.version)?;
-		fill_import_params(&mut config, &self.params.import_params, sc_service::Roles::FULL)?;
-
-		let file: Box<dyn ReadPlusSeek + Send> = match self.params.input {
-			Some(filename) => Box::new(File::open(filename)?),
-			None => {
-				let mut buffer = Vec::new();
-				stdin().read_to_end(&mut buffer)?;
-				Box::new(Cursor::new(buffer))
-			},
-		};
+			match fs::remove_dir_all(&db_path) {
+				Result::Ok(_) => {
+					println!("{:?} removed.", &db_path);
+					Ok(())
+				},
+				Result::Err(ref err) if err.kind() == ErrorKind::NotFound => {
+					eprintln!("{:?} did not exist.", &db_path);
+					Ok(())
+				},
+				Result::Err(err) => Result::Err(err.into())
+			}
+		},
+		CoreParams::Revert(params) => {
+			let config = get_config(&core_params, spec_factory, impl_name, &version)?;
 
-		// Note: while we would like the user to handle the exit themselves, we handle it here
-		// for backwards compatibility reasons.
-		let (exit_send, exit_recv) = std::sync::mpsc::channel();
-		let exit = exit.into_exit();
-		std::thread::spawn(move || {
-			block_on(exit);
-			let _ = exit_send.send(());
-		});
+			let blocks = params.num.parse()?;
+			builder(config)?.revert_chain(blocks)?;
 
-		let mut import_fut = builder(config)?
-			.import_blocks(file, false)
-			.compat();
-		let fut = futures::future::poll_fn(|cx| {
-			if exit_recv.try_recv().is_ok() {
-				return Poll::Ready(Ok(()));
-			}
-			Pin::new(&mut import_fut).poll(cx)
-		});
+			Ok(())
+		},
+	};
 
-		let mut runtime = tokio::runtime::Runtime::new().unwrap();
-		runtime.block_on(fut)?;
-		Ok(())
-	}
+	Ok(())
 }
 
-/// Command to check a block.
-pub struct CheckBlock<'a> {
-	params: CheckBlockCmd,
-	version: &'a VersionInfo,
+pub fn run_node<G, E, F2, F3, T1, T2>(
+	config: Configuration<G, E>,
+	new_light: F2,
+	new_full: F3,
+	version: &VersionInfo,
+) -> error::Result<()>
+where
+	F2: FnOnce(Configuration<G, E>) -> Result<T1, sc_service::error::Error>,
+	F3: FnOnce(Configuration<G, E>) -> Result<T2, sc_service::error::Error>,
+	G: RuntimeGenesis,
+	E: ChainSpecExtension,
+	T1: AbstractService + Unpin,
+	T2: AbstractService + Unpin,
+{
+	info!("{}", version.name);
+	info!("  version {}", config.full_version());
+	info!("  by {}, 2017, 2018", version.author);
+	info!("Chain specification: {}", config.chain_spec.name());
+	info!("Node name: {}", config.name);
+	info!("Roles: {}", display_role(&config));
+	match config.roles {
+		ServiceRoles::LIGHT => run_service_until_exit(
+			new_light(config)?,
+		),
+		_ => run_service_until_exit(
+			new_full(config)?,
+		),
+	}
 }
```

Comment:
> > I believe this change is incompatible with #4688, as we have to create the tokio runtime and modify the `Configuration` before creating the `Service`.
> 
> I think you're right. It's also useful for cumulus because we have to start 2 nodes so we need a better control over the runtime. I will try to improve this.

@tomaka I'm making a thread here in the code to make it easier to track on GitHub.

This is what I had in mind to solve this problem: offer the user to match the subcommand with RunCmd so you can add a match branch for your custom implementation and let it fallback to the default for the rest.

Here is an example (based on bin/node-template/cli):
```rust
	match &core_params {
		CoreParams::RunCmd(cmd_params) => {
			let config = get_config(....);
			total_control_on_how_you_run_your_node();
		},
		params => sc_cli::run(
			params,
			service::new_light,
			service::new_full,
			load_spec,
			|config: Config<_, _>| Ok(new_full_start!(config).0),
			"substrate-node",
			&version,
		},
	}
```

So for RunCmd it will run your own tokio runtime while for the other commands it will use the defaults.

Reply:
The problem to me, is: this `RunCmd` situation is one example of one of the issues that could arise from hiding the behaviour inside of that `sc_cli::run` function.

The reason why I'm not a fan of this specific change is that we're locking ourselves inside of that specific use case, and if we ever need more control then we will start piling up hacks. From my experience, having control over what happens taken away from users is exactly how a source code starts having hacks.

Reply:
Do you call this a hack?
```rust
match args {
    Variant1 => process1(),
    Variant2 => process2(),
    other => catch_all(),
}
```

I see `sc_cli::run` as a helper function for a catch-all argument of a match arm. If we don't do that, every CLI must implements every variant of substrate even if they don't give a care.

I think if you compare this:
```rust
	match subcommand {
		Cli::SubstrateCli(cli) => sc_cli::run(
			cli,
			service::new_light,
			service::new_full,
			load_spec,
			|config: Config<_, _>| Ok(new_full_start!(config).0),
			"substrate-node",
			&version,
		),
		Cli::Factory(cli_args) => {
			...
		},
	}
```
vs that:
```rust
	match parse_and_prepare::<CustomSubcommands, NoCustom, _>(&version, "substrate-node", args) {
		ParseAndPrepare::Run(cmd) => cmd.run(load_spec, exit,
		|exit, _cli_args, _custom_args, config: Config<_, _>| {
			info!("{}", version.name);
			info!("  version {}", config.full_version());
			info!("  by Parity Technologies, 2017-2019");
			info!("Chain specification: {}", config.chain_spec.name());
			info!("Node name: {}", config.name);
			info!("Roles: {}", display_role(&config));
			let runtime = RuntimeBuilder::new().name_prefix("main-tokio-").build()
				.map_err(|e| format!("{:?}", e))?;
			match config.roles {
				ServiceRoles::LIGHT => run_until_exit(
					runtime,
					service::new_light(config)?,
					exit
				),
				_ => run_until_exit(
					runtime,
					service::new_full(config)?,
					exit
				),
			}
		}),
		ParseAndPrepare::BuildSpec(cmd) => cmd.run::<NoCustom, _, _, _>(load_spec),
		ParseAndPrepare::ExportBlocks(cmd) => cmd.run_with_builder(|config: Config<_, _>|
			Ok(new_full_start!(config).0), load_spec, exit),
		ParseAndPrepare::ImportBlocks(cmd) => cmd.run_with_builder(|config: Config<_, _>|
			Ok(new_full_start!(config).0), load_spec, exit),
		ParseAndPrepare::CheckBlock(cmd) => cmd.run_with_builder(|config: Config<_, _>|
			Ok(new_full_start!(config).0), load_spec, exit),
		ParseAndPrepare::PurgeChain(cmd) => cmd.run(load_spec),
		ParseAndPrepare::RevertChain(cmd) => cmd.run_with_builder(|config: Config<_, _>|
			Ok(new_full_start!(config).0), load_spec),
		ParseAndPrepare::CustomCommand(CustomSubcommands::Factory(cli_args)) => {
			...
		}
	}
```

You can clearly see it became easier to use substrate. I'm skipping the types but you can see the number of generics needed for the custom command, this is also a big part of the benefit. I'm also skipping the IntoExit that must be implemented but here it is provided automatically if you don't need something special.

And most of the steps inside `sc_cli::run` can be imported separately if needed (the signal handling and `get_config`)

Reply:
I think we should go use some sort of mixed solution here.

- Create a new crate "opinionated-cli" (Find a better name :D)
- This new crates does all these repetitive stuff like `match parse()` blah.
- Polkadot/Substrate node can use this.
- If someone wants to go the extra mile, they can just use `sc-cli` and do it by hand. 

TLDR: Keep the interface of `sc-cli` as generic as possible, but move all the repetitive/common stuff into own crate.

Reply:
Your code is not a hack. I'm saying that your code will lead us to being locked in a specific use case, which may cause problems in the future if something doesn't fit exactly in one of the three provided ways to create a service.

You're also comparing apples vs oranges. I'm obviously in favour of removing `IntoExit` and all. My grief is only about this `run` function.

I know it's a bad argumentation, but I'm mostly drawing this out of experience because this `run` function looks like the kind of code that usually causes the most trouble a couple months later.

Reply:
> TLDR: Keep the interface of sc-cli as generic as possible, but move all the repetitive/common stuff into own crate.

That's fine by me :woman_shrugging: 

> I'm saying that your code will lead us to being locked in a specific use case, which may cause problems in the future if something doesn't fit exactly in one of the three provided ways to create a service.

I trust your experience on this don't worry. I'm just trying to understand and find solutions. I guess I don't 100% understand your point yet. This PR fixes most of the issues/limitations I encountered when I worked on cumulus CLI (as a user of substrate), it's really an attempt to make it also more flexible than what it is right now. That is why I'm exposing RunCmd now (which wasn't exposed before).

Does the solution of @bkchr sound good? I don't think it would be bad to keep client/cli as basic/simple as possible and provide opinionated helpers in another module (which would provide signal handling and running most of the commands). (Or to its own crate)

What do you think?

Reply:
For me this sounds okay and as said before, I would move the opinionated into its own crate. To have a better distinction.

Reply:
I like the solution for the fact that we're giving the choice.

> Create a new crate "opinionated-cli" (Find a better name :D)

But why move the opinionated stuff into a separate crate? Why not put it in `sc-cli`?

I think it's great if `sc-cli` exposes both `parse_and_execute` and `run`, and you can choose which API you want.

Reply:
Yeah or that way, I thought that it would make the distinction more clear.

Reply:
Discarding my concerns after discussing in DM.
The concern is valid, but it's not as bad as I thought and there's no point in trying to have a stable API of sc-cli yet.

Reply:
I added 2 important things:

1. A change on the way Configuration is created. It can be now created with `Configuration::default` which can then be passed in argument to the run functions. The idea is to solve the problem https://github.com/paritytech/cumulus/issues/24 by allowing the user to change the base configuration **before** the command line arguments gets loaded.

Turns out that after doing that, a lot of constraints in the code disappeared and I was able to remove a lot of line of codes again.

The only new constraint added here is that `chain_spec` on Configuration is now an `Option`.

2. I moved the different command actions to their own function again like it was with `ParseAndPrepare`. The only difference is that I use the structopt directly (which is simpler and allows the user to do the exact same thing than ParseAndPrepare before).